src/functions.php

```diff
@@ -478,23 +478,48 @@ function is_settled(PromiseInterface $promise)
 function coroutine(callable $generatorFn)
 {
     $generator = $generatorFn();
-    return __next_coroutine($generator->current(), $generator)->then();
+    $state = new \stdClass;
```

Comment:
Thanks for taking a look. 

> I would suggest creating creating a class to maintain coroutine state. You could even go so far as to encapsulate coroutine internals in a class that implements PromisorInterface.

Creating a new class to encapsulate _all_ coroutine functionality might make sense, although I don't have time to do it, but creating a class simply to hold the internal coroutine state would be wrong imo. What value would be added to justify adding a new class to the API and having to maintain it? What use case would the class have outside of this specific piece of code? Have you thought about the performance impact of all the added function calls?

Unfortunately, I don't have any more time to put into this, but I'm really much more concerned about the bug being fixed than implementation details, so please feel free to close this PR and fix the bug in your own way if you prefer.